common/Renderers/DevHomeActionSet.cs

```diff
@@ -0,0 +1,137 @@
+    /// </summary>
+    private readonly TopLevelCardActionSetVisibility _actionSetVisibility;
+
+    /// <summary>
+    /// Gets the adaptive card object that will invoke an action within the action set.
+    /// </summary>
+    public AdaptiveActionInvoker? ActionButtonInvoker { get; private set; }
+
+    public AdaptiveCard? OriginalAdaptiveCard { get; private set; }
+
+    private Dictionary<string, IAdaptiveActionElement> ActionButtonMap { get; } = new();
+
+    public DevHomeActionSet(TopLevelCardActionSetVisibility cardActionSetVisibility)
+    {
+        _actionSetVisibility = cardActionSetVisibility;
+    }
+
+    public UIElement? Render(IAdaptiveCardElement element, AdaptiveRenderContext context, AdaptiveRenderArgs renderArgs)
+    {
+        ActionButtonMap.Clear();
+
+        if (element is AdaptiveActionSet actionSet)
+        {
+            foreach (var action in actionSet.Actions)
+            {
+                if (action is AdaptiveExecuteAction executeAction)
+                {
+                    context.LinkSubmitActionToCard(executeAction, renderArgs);
+                }
+                else if (action is AdaptiveSubmitAction submitAction)
+                {
+                    context.LinkSubmitActionToCard(submitAction, renderArgs);
+                }
+
+                ActionButtonMap.TryAdd(action.Id, action);
+            }
+
+            ActionButtonInvoker = context.ActionInvoker;
+            OriginalAdaptiveCard = renderArgs.ParentCard;
+
+            if (_actionSetVisibility == TopLevelCardActionSetVisibility.Hidden)
+            {
+                // the page in Dev Home does not want to show the action set in the card.
+                // So we return null to prevent the adaptive card buttons from appearing.
+                // Invoking the button from Dev Home will then trigger the action in the adaptive card.
+                return null;
+            }
+        }
+
+        var renderer = new AdaptiveActionSetRenderer();
+        return renderer.Render(element, context, renderArgs);
+    }
+
+    /// <summary>
+    /// Invokes an adaptive card action from anywhere within Dev Home, like a method in a view Model for example.
+    /// A boolean is returned with the validation result of the card. We still send the action event so the adaptive
+    /// cards UI updates with error information.
+    /// </summary>
+    /// <returns>
+    /// A boolean indicating whether validation for the card passed or failed.
+    /// </returns>
+    public bool TryValidateAndInitiateAction(string buttonId, AdaptiveInputs userInputs)
+    {
+        ActionButtonMap.TryGetValue(buttonId, out var actionElement);
+
+        if (actionElement == null)
+        {
+            return false;
+        }
+
+        var result = true;
+        if (!userInputs.ValidateInputs(actionElement))
+        {
+            result = false;
+        }
+
+        ActionButtonInvoker?.SendActionEvent(actionElement);
```

Comment:
If result is false, do we not want to continue?

Reply:
If result is false, its up to who-ever initiated the call to decide what to do. In the case of the machine configuration flow, it won't proceed to the next page. We still do the `ActionButtonInvoker?.SendActionEvent(actionElement);` to submit the action to the adaptive card,  so the general adaptive card error UI for the control shows up in the UI. For example, in the case of a Text.Input type in an adaptive card with an "isRequired" property. If we submit the action and the validation result is false, the adaptive card code will highlight the textbox in red for us in the UI, without Dev Home needing to do anything. So that's why I still send the action when its false.